File: android/app/src/main/kotlin/dev/andromac/core/Session.kt
package dev.andromac.core

import org.json.JSONObject
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.EOFException
import java.io.IOException
import java.net.Socket
import java.security.PrivateKey
import java.security.PublicKey

/**
 * The handshake verified cryptographically (the peer proved possession of its static private
 * key), but this device is not paired yet, or the pinned key has changed.
 */
class UntrustedPeerException(
    val peerStaticPub: ByteArray,
    val peerName: String,
    val sas: String,
    /** true means: a key was pinned before and it has CHANGED — a serious warning. */
    val pinMismatch: Boolean,
) : IOException(if (pinMismatch) "pinned key changed" else "device not paired")

/**
 * A docs/PROTOCOL.md §2/§4 session running over an already established TCP socket.
 * Android is always the initiator (role A).
 */
class Session internal constructor(
    private val socket: Socket,
    private val input: DataInputStream,
    private val output: DataOutputStream,
    private val sendKey: ByteArray,
    private val recvKey: ByteArray,
    val peerStaticPub: ByteArray,
    val sas: String,
) : AutoCloseable {

    private var sendCounter = 1L      // 0 was consumed by the confirmation round
    private var recvCounter = 1L

    @Synchronized
    fun send(msg: JSONObject) {
        val ct = Crypto.seal(sendKey, sendCounter++, msg.toString().toByteArray(Charsets.UTF_8))
        writeFrame(output, ct)
    }

    /** Reads and decrypts one message. The socket read timeout is the caller's responsibility. */
    fun receive(): JSONObject {
        val ct = readFrame(input)
        val pt = Crypto.open(recvKey, recvCounter++, ct)   // an AEAD failure means: drop the link
        return JSONObject(String(pt, Charsets.UTF_8))
    }

    override fun close() {
        // Deliberately NOT under the monitor: closing the socket is what unblocks a [send]
        // stuck writing to a peer that stopped reading. Taking the lock first would let a
        // stalled write hang whoever closes the link, including the main thread in onDestroy.
        runCatching { socket.close() }
        zeroKeys()
    }

    /**
     * Best-effort zeroing of the session keys so they do not linger in a heap dump or in swap
     * after close. Not a guarantee (the GC may have copied them), but it costs nothing.
     *
     * Synchronized against [send]: zeroing while a queued send sat inside `Crypto.seal` used
     * to produce a frame encrypted under a half-zeroed key. The socket is already closed by
     * the time we get here, so the in-flight write cannot block us for long.
     */
    @Synchronized
    private fun zeroKeys() {
        java.util.Arrays.fill(sendKey, 0)
        java.util.Arrays.fill(recvKey, 0)
    }

    companion object {
        const val MAX_FRAME = 1 shl 20      // 1 MiB, PROTOCOL §4
        private const val PROTO: Byte = 3   // handshake version byte, PROTOCOL §2
        private val LABEL = "AndroMac/v3".toByteArray(Charsets.UTF_8)

        private fun writeFrame(out: DataOutputStream, payload: ByteArray) {
            out.writeInt(payload.size)
            out.write(payload)
            out.flush()
        }

        private fun readFrame(input: DataInputStream): ByteArray {
            val n = input.readInt()
            if (n <= 0 || n > MAX_FRAME) throw IOException("invalid frame length: $n")
            return ByteArray(n).also(input::readFully)
        }

        /**
         * Client-side (role A) handshake.
         *
         * [pinnedKeyFor] returns the stored pin if there is one, otherwise null. A pinned key
         * that does not match ends the attempt before this phone's static key is sent
         * ([UntrustedPeerException] with `pinMismatch = true` and no SAS). With no pin, the
         * handshake completes, the peer proves its key, and the exception carries the SAS to
         * show; the socket is closed either way.
         */
        @Throws(IOException::class)
        fun connect(
            socket: Socket,
            staticPriv: PrivateKey,
            staticPubBytes: ByteArray,
            peerName: String,
            pinnedKeyFor: () -> ByteArray?,
        ): Session {
            val input = DataInputStream(socket.getInputStream().buffered())
            val output = DataOutputStream(socket.getOutputStream().buffered())

            // 1. A -> B — only the ephemeral key; the static key stays hidden until the Mac has
            //    proven (via dh3) to be the one we pinned, so a passive listener or a stranger
            //    that merely answers on the port never learns this phone's identity key.
            val eph = Crypto.generateKeyPair()
            val ePubA = Crypto.encodePublic(eph.public)
            output.write(byteArrayOf(PROTO))
            output.write(ePubA)
            output.flush()

            // 2. B -> A — e_B in the clear, then s_B and the nonce commitment c_B encrypted under
            //    the ephemeral-ephemeral secret. B commits to n_B before it sees n_A.
            val proto = input.read()
            if (proto == -1) throw EOFException("connection dropped during the handshake")
            if (proto != PROTO.toInt()) throw IOException("unsupported protocol version: $proto")
            val ePubB = ByteArray(Crypto.PUB_LEN).also(input::readFully)
            val ctB = ByteArray(Crypto.PUB_LEN + Crypto.NONCE_LEN + Crypto.TAG_LEN).also(input::readFully)

            val peerEph: PublicKey = Crypto.decodePublic(ePubB)   // on-curve validation happens here
            val h1 = Crypto.sha256(LABEL, ePubA, ePubB)
            val dh1 = Crypto.ecdh(eph.private, peerEph)           // forward secrecy
            val ptB = try {
                Crypto.open(Crypto.hkdf(h1, dh1, "AndroMac/v3 ee", 32), 0, ctB)
            } catch (e: Exception) {
                socket.close()
                throw IOException("message 2 failed to open — key agreement mismatch", e)
            }
            val sPubB = ptB.copyOfRange(0, Crypto.PUB_LEN)
            val cB = ptB.copyOfRange(Crypto.PUB_LEN, ptB.size)
            val peerStatic: PublicKey = Crypto.decodePublic(sPubB)

            // The pin is checked BEFORE our static key goes out: a changed key ends the attempt
            // here, with nothing about this phone revealed. There is no SAS yet — the user
            // resets the pairing and the next attempt shows a fresh code.
            val pinned = pinnedKeyFor()
            if (pinned != null && !Crypto.constantTimeEquals(pinned, sPubB)) {
                socket.close()
                throw UntrustedPeerException(sPubB, peerName, sas = "", pinMismatch = true)
            }

            // 3. A -> B — s_A encrypted so that only the holder of s_B's private key can read it.
            val dh3 = Crypto.ecdh(eph.private, peerStatic)        // identity of B
            val ctA = Crypto.seal(Crypto.hkdf(h1, dh1 + dh3, "AndroMac/v3 es", 32), 0, staticPubBytes)
            output.write(ctA)

            val transcript = Crypto.sha256(LABEL, ePubA, ePubB, ctB, ctA)
            val ikm = dh1 +
                Crypto.ecdh(staticPriv, peerEph) +             // dh2: identity of A
                dh3
            val sendKey = Crypto.hkdf(transcript, ikm, "AndroMac a2b", 32)
            val recvKey = Crypto.hkdf(transcript, ikm, "AndroMac b2a", 32)

            // 3./4. Confirmation round — nonce 0, plaintext = transcript || own nonce. Sent in
            //    the same flush as ct_A.
            val nA = Crypto.randomBytes(Crypto.NONCE_LEN)
            writeFrame(output, Crypto.seal(sendKey, 0, transcript + nA))
            val theirs = try {
                Crypto.open(recvKey, 0, readFrame(input))
            } catch (e: Exception) {
                socket.close()
                throw IOException("confirmation round failed — key agreement mismatch or MITM", e)
            }
            if (theirs.size != transcript.size + Crypto.NONCE_LEN ||
                !Crypto.constantTimeEquals(theirs.copyOfRange(0, transcript.size), transcript)
            ) {
                socket.close()
                throw IOException("transcript mismatch — suspected MITM")
            }
            val nB = theirs.copyOfRange(transcript.size, theirs.size)
            if (!Crypto.constantTimeEquals(Crypto.commit(nB), cB)) {
                socket.close()
                throw IOException("nonce commitment mismatch — suspected MITM")
            }

            // Past this point the peer has proven possession of the s_B private key.
            val sas = Crypto.sas(transcript, nA, nB)
            if (pinned == null) {
                socket.close()
                throw UntrustedPeerException(sPubB, peerName, sas, pinMismatch = false)
            }

            return Session(socket, input, output, sendKey, recvKey, sPubB, sas)
        }
    }
}

File: macos/Sources/AndroMacKit/Session.swift
import CryptoKit
import Foundation
import Network

/// An encrypted session over an established connection. macOS is always the responder (role B).
public actor Session {

    private let wire: Wire
    private let sendKey: SymmetricKey
    private let recvKey: SymmetricKey
    private var sendCounter: UInt64 = 1      // 0 was spent on the verification round
    private var recvCounter: UInt64 = 1

    public nonisolated let peerStaticPub: Data
    public nonisolated let sas: String

    private init(wire: Wire, sendKey: SymmetricKey, recvKey: SymmetricKey, peerStaticPub: Data, sas: String) {
        self.wire = wire
        self.sendKey = sendKey
        self.recvKey = recvKey
        self.peerStaticPub = peerStaticPub
        self.sas = sas
    }

    /// Returns the plaintext size in bytes (for the traffic counters; the content is never logged).
    @discardableResult
    public func send(_ message: [String: Any]) async throws -> Int {
        try await send(json: JSONSerialization.data(
            withJSONObject: message, options: [.withoutEscapingSlashes]
        ))
    }

    /// Send a body that is already JSON.
    ///
    /// This is what lets the Mac broadcast to several phones: the message is encoded once and then
    /// sealed once per session, because each session has its own key and its own counter. Handing
    /// the same `[String: Any]` to N sessions instead would be both wasteful and, since `Any` can
    /// hold a reference, something the compiler is right to refuse.
    public func send(json: Data) async throws -> Int {
        let ct = try Crypto.seal(sendKey, sendCounter, json)
        sendCounter &+= 1
        try await wire.writeFrame(ct)
        return json.count
    }

    public func receive() async throws -> sending [String: Any] {
        try await receiveCounted().message
    }

    /// Like `receive()`, plus the plaintext size in bytes.
    public func receiveCounted() async throws -> sending (message: [String: Any], bytes: Int) {
        let ct = try await wire.readFrame()
        let pt = try Crypto.open(recvKey, recvCounter, ct)
        recvCounter &+= 1
        guard let obj = try JSONSerialization.jsonObject(with: pt) as? [String: Any] else {
            throw WireError.protocolError("expected a JSON object")
        }
        return (obj, pt.count)
    }

    public func close() async { await wire.close() }

    // MARK: handshake (role B)

    /// Handshake version byte (docs/PROTOCOL.md §2).
    static let proto: UInt8 = 3
    static let label = Data("AndroMac/v3".utf8)

    /// docs/PROTOCOL.md §2. If the peer's static key is not among `pinnedKeys`, throws
    /// `WireError.untrusted` and the connection is closed — the caller can then show the SAS to the
    /// user and pin it.
    ///
    /// The Mac trusts a *set* of phones, so this is a membership test rather than one comparison.
    /// Note what that changes: there is no longer such a thing as "the pinned key has changed".
    /// A key we do not know is simply a device we have not met, whether it is a new phone or a
    /// phone that was wiped and reinstalled — and either way the only safe answer is the same, a
    /// fresh SAS comparison. What protects the user is the approval, never the label on it.
    public static func accept(
        connection: NWConnection,
        queue: DispatchQueue,
        staticKey: P256.KeyAgreement.PrivateKey,
        peerName: String,
        pinnedKeys: [Data]
    ) async throws -> Session {
        let wire = Wire(connection)

        do {
            // A connection that never reaches .ready must be closed too: when start() sat
            // OUTSIDE the do block, an error left the NWConnection open.
            try await wire.start(on: queue)

            // 1. A -> B — the phone's ephemeral key only; its static key arrives in message 3,
            //    encrypted so that only the holder of our static private key can read it.
            let head = try await wire.receive(exactly: 1 + Crypto.pubLen)
            guard head[head.startIndex] == Self.proto else {
                throw WireError.protocolError("unsupported protocol version: \(head[head.startIndex])")
            }
            let ePubA = Data(head.suffix(Crypto.pubLen))
            let peerEph = try Crypto.decodePublic(ePubA)

            // 2. B -> A — e_B in the clear; s_B and the commitment c_B under the ephemeral-
            //    ephemeral secret. Commit to n_B before A reveals n_A, so neither side can
            //    steer the SAS.
            let sPubBData = Crypto.encodePublic(staticKey.publicKey)
            let eph = Crypto.generateKeyPair()
            let ePubBData = Crypto.encodePublic(eph.publicKey)
            let nB = Crypto.randomBytes(Crypto.nonceLen)
            let cB = Crypto.commit(nB)
            let h1 = Crypto.sha256(Self.label, ePubA, ePubBData)
            // CAUTION: the DH names are written from the initiator's (A) point of view; on B the
            // peer roles are reversed.
            //   dh1 = ECDH(e_A, e_B)  ->  here ECDH(e_B, e_A)
            //   dh2 = ECDH(s_A, e_B)  ->  here ECDH(e_B, s_A)
            //   dh3 = ECDH(e_A, s_B)  ->  here ECDH(s_B, e_A)
            let dh1 = try Crypto.ecdh(eph, peerEph)              // forward secrecy
            let dh3 = try Crypto.ecdh(staticKey, peerEph)        // B's identity
            let ctB = try Crypto.seal(Crypto.hkdf(salt: h1, ikm: dh1, info: "AndroMac/v3 ee", count: 32), 0, sPubBData + cB)
            try await wire.send(Data([Self.proto]) + ePubBData + ctB)

            // 3. A -> B — s_A, readable only with dh3. A phone that pinned a different key
            //    hangs up instead of sending this.
            let ctA = try await wire.receive(exactly: Crypto.pubLen + Crypto.tagLen)
            let sPubA: Data
            do {
                sPubA = try Crypto.open(Crypto.hkdf(salt: h1, ikm: dh1 + dh3, info: "AndroMac/v3 es", count: 32), 0, ctA)
            } catch {
                throw WireError.protocolError("message 3 failed to open — no key agreement, or a MITM")
            }
            let peerStatic = try Crypto.decodePublic(sPubA)

            let transcript = Crypto.sha256(Self.label, ePubA, ePubBData, ctB, ctA)
            let ikm = dh1
                + (try Crypto.ecdh(eph, peerStatic))             // dh2: A's identity
                + dh3

            let kA2B = Crypto.hkdf(salt: transcript, ikm: ikm, info: "AndroMac a2b", count: 32)
            let kB2A = Crypto.hkdf(salt: transcript, ikm: ikm, info: "AndroMac b2a", count: 32)

            // 3b. A -> B verification (transcript || n_A), 4. B -> A verification (transcript || n_B)
            let theirs: Data
            do {
                theirs = try Crypto.open(kA2B, 0, try await wire.readFrame())
            } catch {
                throw WireError.protocolError("verification round failed — no key agreement, or a MITM")
            }
            guard theirs.count == transcript.count + Crypto.nonceLen,
                  Crypto.constantTimeEquals(theirs.prefix(transcript.count), transcript) else {
                throw WireError.protocolError("transcript mismatch — MITM suspected")
            }
            let nA = Data(theirs.suffix(Crypto.nonceLen))
            try await wire.writeFrame(try Crypto.seal(kB2A, 0, transcript + nB))

            // From here on the peer has proven that it holds the s_A private key.
            let sas = Crypto.sas(transcript: transcript, nA: nA, nB: nB)
            // No early exit: every pinned key is compared, so the loop takes the same time whether
            // the match is the first entry or the last. The keys are public, so this is hygiene
            // rather than a defence, and it is cheap enough to keep.
            var trusted = false
            for pinned in pinnedKeys where Crypto.constantTimeEquals(pinned, sPubA) { trusted = true }
            guard trusted else {
                throw WireError.untrusted(
                    peerKey: sPubA, peerName: peerName, sas: sas, isFirstDevice: pinnedKeys.isEmpty
                )
            }

            return Session(wire: wire, sendKey: kB2A, recvKey: kA2B, peerStaticPub: sPubA, sas: sas)
        } catch {
            await wire.close()
            throw error
        }
    }
}
